### PR TITLE
Make the static state of ToolLocationHelper thread safe

### PR DESCRIPTION
**Pull request by cdmihai**
Some fields were not protected by the lock. Added the lock.

This branch was made out of the update 2 branch, therefore this PR is also a merge of update2 into master

### CODE REVIEW

**Comment by rainersigwald** [comment]
Can you reference #438 in the commit that you fixed it with? (I'm fighting to never be too lazy to _file_ the bug, even if I'm too lazy to fix it)

**Comment by rainersigwald** [comment]
Made a lightweight audit of the static caches and they all appear to be under lock now :+1: 

Do you think it's worth filing an issue to change them all to ConcurrentDictionaries like we discussed?

LGTM

**Comment by cdmihai** [comment]
Added the #438 to the commit message.

Regarding the concurrent collections, do you have a sense of how often is the tool location helper called? Also, from what I saw, all those caches contain a very well defined and finite set of possible values. If the contention is not high (e.g. one thread calls into it once per build) then I do not think it's worth it.

**Comment by rainersigwald** [comment]
I doubt contention is particularly significant and I'm willing for profiling to show otherwise. I suspect it's on the order of ~10 calls per project, and these lookups should go pretty quickly after the initial population, so I'm fine with doing nothing else now.

File: src/XMakeTasks/UnitTests/GetReferencePaths_Tests.cs [hunk 1]
```diff
@@ -157,16 +157,51 @@ public void TestGeneralFrameworkMonikerNonExistent()
             MockEngine engine = new MockEngine();
             GetReferenceAssemblyPaths getReferencePaths = new GetReferenceAssemblyPaths();
             getReferencePaths.BuildEngine = engine;
-            // Make a framework which does not exist, intentional mispelling of framework
+            // Make a framework which does not exist, intentional misspelling of framework
             getReferencePaths.TargetFrameworkMoniker = ".NetFramewok, Version=v99.0";
-            getReferencePaths.Execute();
+            bool success = getReferencePaths.Execute();
```
Comment (kivancmuslu) [inline src/XMakeTasks/UnitTests/GetReferencePaths_Tests.cs]: This is a test comment, please ignore.

File: src/XMakeTasks/UnitTests/GetReferencePaths_Tests.cs [hunk 2]
```diff
@@ -157,16 +157,51 @@ public void TestGeneralFrameworkMonikerNonExistent()
             MockEngine engine = new MockEngine();
```
Comment (kivancmuslu) [inline src/XMakeTasks/UnitTests/GetReferencePaths_Tests.cs]: Another test comment, please ignore.

File: src/Utilities/ToolLocationHelper.cs [hunk 3]
```diff
@@ -853,30 +853,33 @@ public static string[] GetTargetPlatformReferences
                 string registryRoot
             )
         {
-            if (s_cachedTargetPlatformReferences == null)
+            lock (s_locker)
```
Comment (kivancmuslu) [inline src/Utilities/ToolLocationHelper.cs]: Just a test comment, please ignore.

**Comment by AndyGerlicher** [comment]
LGTM

**Pull request merged**